pkg/spoke/helper/helper.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	workapiv1 "open-cluster-management.io/api/work/v1"
+	"open-cluster-management.io/work/pkg/helper"
+	"open-cluster-management.io/work/pkg/spoke/auth"
+)
+
+// DeleteAppliedResources deletes all given applied resources and returns those pending for finalization
+// If the uid recorded in resources is different from what we get by client, ignore the deletion.
+func DeleteAppliedResources(
+	ctx context.Context,
+	resources []workapiv1.AppliedManifestResourceMeta,
+	reason string,
+	dynamicClient dynamic.Interface,
+	recorder events.Recorder,
+	owner metav1.OwnerReference,
+	kubeClient kubernetes.Interface, executor *workapiv1.ManifestWorkExecutor) ([]workapiv1.AppliedManifestResourceMeta, []error) {
+	var resourcesPendingFinalization []workapiv1.AppliedManifestResourceMeta
+	var errs []error
+
+	// set owner to be removed
+	ownerCopy := owner.DeepCopy()
+	ownerCopy.UID = types.UID(fmt.Sprintf("%s-", owner.UID))
+
+	// We hard coded the delete policy to Background
+	// TODO: reivist if user needs to set other options. Setting to Orphan may not make sense, since when
+	// the manifestwork is removed, there is no way to track the orphaned resource any more.
+	deletePolicy := metav1.DeletePropagationBackground
+
+	validator := auth.NewExecutorValidator(kubeClient)
+
+	for _, resource := range resources {
+		gvr := schema.GroupVersionResource{Group: resource.Group, Version: resource.Version, Resource: resource.Resource}
+		u, err := dynamicClient.
+			Resource(gvr).
+			Namespace(resource.Namespace).
+			Get(ctx, resource.Name, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			klog.V(2).Infof("Resource %v with key %s/%s is removed Successfully", gvr, resource.Namespace, resource.Name)
+			continue
+		}
+
+		if err != nil {
+			errs = append(errs, fmt.Errorf(
+				"failed to get resource %v with key %s/%s: %w",
+				gvr, resource.Namespace, resource.Name, err))
+			continue
+		}
+
+		existingOwner := u.GetOwnerReferences()
+
+		// If it is not owned by us, skip
+		if !helper.IsOwnedBy(owner, existingOwner) {
+			continue
+		}
+
+		err = validator.ValidateDelete(ctx, executor, gvr, resource.Namespace, resource.Name)
+		if err != nil {
+			if !apierrors.IsForbidden(err) {
+				errs = append(errs, fmt.Errorf(
+					"failed to check delete permission for resource %v with key %s/%s: %w",
+					gvr, resource.Namespace, resource.Name, err))
+				continue
+			}
+			// TODO: consider to reflect the not allowed error on the manifest work condition
```

Comment:
this will be certainly needed. I actually think all delete action should be put in the separate PR together with how we handle condition.